drivers/i2s/i2s_nrfx.c

```diff
@@ -0,0 +1,1336 @@
+		 channel_change_state(channel_to_drain, I2S_STATE_STOPPING);
+}
+
+static bool channel_check_empty(struct channel_str *channel)
+{
+	return queue_is_empty(&channel->mem_block_queue);
+}
+
+static int channel_tx_get_data(struct i2s_nrfx_data *i2s, u32_t **buf)
+{
+	struct channel_str *ch_tx = &i2s->channel_tx;
+	int ret;
+
+	assert(i2s != NULL);
+	ret = queue_fetch(&ch_tx->mem_block_queue, (void **)buf);
+	if (ret < 0) {
+		return ret;
+	}
+	k_sem_give(&ch_tx->sem);
+	return 0;
+}
+
+/* In case of constant tx transmission this callback:
+ * - frees tx buffer which has just been transmitted via I2S interface
+ * - gets new buffer from queue which will be passed to nrfx driver as
+ *   next to be transferred (content of 'TXD.PTR' register).
+ *   Status parameter informs about handler execution reason:
+ *   - if 1 - next buffer is needed
+ *   - if 0 - transfer is finishing
+ */
+static void channel_tx_callback(struct i2s_nrfx_data *i2s,
+			nrfx_i2s_buffers_t const *p_released,
+			u32_t status, nrfx_i2s_buffers_t *p_new_buffers)
+{
+	struct channel_str *ch_tx = &i2s->channel_tx;
+	u32_t *mem_block = NULL;
+	int get_data_ret = 1;
+
+	if (ch_tx->current_state == I2S_STATE_RUNNING &&
+	    interface_get_state(i2s) == I2S_IF_NEEDS_RESTART) {
+		/* This code services case when tx channel transmits data
+		 * constantly while rx channel is beginning/finishing its
+		 * transfer (user called' i2s_trigger()' with 'I2S_DIR_RX'
+		 * parameter)).
+		 * In this case NRF I2S peripheral needs to be restarted.
+		 */
+		if (p_released->p_tx_buffer != NULL &&
+		    next_buffers_needed(status)) {
+			/* when interface needs to be restarted and
+			 * last event was 'EVENT_STOPPED' then we don't
+			 * free this buffer - it will be used after
+			 * after interface restarts - tx transmission
+			 * will be still running and user don't want to
+			 * lose data.
+			 */
+			k_mem_slab_free(ch_tx->mem_slab,
+					(void **)&p_released->p_tx_buffer);
+		}
+		get_data_ret = channel_tx_get_data(i2s, &mem_block);
+		if (get_data_ret == 0) {
+			k_mem_slab_free(ch_tx->mem_slab,
+					(void **)&mem_block);
+		}
+		return;
+	}
+	if (p_released->p_tx_buffer != NULL) {
+		k_mem_slab_free(ch_tx->mem_slab,
+				(void **)&p_released->p_tx_buffer);
+	}
+	if (next_buffers_needed(status)) {
+		get_data_ret = channel_tx_get_data(i2s, &mem_block);
+	}
+	if (ch_tx->current_state == I2S_STATE_STOPPING) {
+		/* finishing tx transfer caused by user trigger command*/
+		enum i2s_trigger_cmd ch_cmd = ch_tx->last_trigger_cmd;
+
+		if (next_buffers_needed(status)) {
+			switch (ch_cmd) {
+			case I2S_TRIGGER_DROP:
+				if (get_data_ret == 0) {
+					channel_tx_mem_clear(i2s, mem_block);
+				}
+				break;
+			case I2S_TRIGGER_STOP:
+				if (get_data_ret == 0) {
+					k_mem_slab_free(ch_tx->mem_slab,
+							(void **)&mem_block);
+				}
+				break;
+			case I2S_TRIGGER_DRAIN:
+				break;
+			default:
+				channel_change_state(ch_tx, I2S_STATE_ERROR);
+				return;
```

Comment:
This part of code needs to be described as well, and perhaps reworked. Currently, it is too hard to understand what is actually going on here. For instance, what happens with `mem_block` here (when the error is signaled)? When will it get freed?